src/browser/components/SettingsPage.jsx

```diff
@@ -65,99 +47,124 @@ export default class SettingsPage extends React.Component {
     this.handleShowUnreadBadge = this.handleShowUnreadBadge.bind(this);
     this.handleChangeUseSpellChecker = this.handleChangeUseSpellChecker.bind(this);
     this.handleChangeEnableHardwareAcceleration = this.handleChangeEnableHardwareAcceleration.bind(this);
+    this.saveSetting = this.saveSetting.bind(this);
+    this.updateSaveState = this.updateSaveState.bind(this);
     this.updateTeam = this.updateTeam.bind(this);
     this.addServer = this.addServer.bind(this);
 
     this.trayIconThemeRef = React.createRef();
+
+    this.saveQueue = [];
   }
 
   componentDidMount() {
+    config.on('update', (configData) => {
+      this.updateSaveState();
+      this.setState(this.convertConfigDataToState(configData, this.state));
+    });
+
+    config.on('error', (error) => {
+      console.log('Config saving error: ', error);
+
+      const savingState = Object.assign({}, this.state.savingState);
+      Object.entries(savingState).forEach(([configType, currentState]) => {
+        if (currentState !== AutoSaveIndicator.SAVING_STATE_DONE) {
+          savingState[configType] = AutoSaveIndicator.SAVING_STATE_ERROR;
+          this.setState({savingState});
+        }
+      });
+    });
+
+    config.on('synchronize', () => {
+      ipcRenderer.send('reload-config');
```

Comment:
Is this calling the event handler below? If that is the case why not just call `config.reload()` instead of another event

Reply:
This is the part of the architecture that I'm not overly fond of as it is somewhat confusing. I have added inline comments to try to explain.

There are two processes running in the desktop app: the main process is run in Node.js and the renderer processes is run in Chromium (kinda like server<->client in web). The first `config.on('synchronize', () => {` function in both locations is used to tell the other process that an update occurred in this process and the second `[ipcMain|ipcRenderer].on('reload-config', () => {` function listens for changes to happen in the other process in order to reload config in this process so it can reload to keep in sync.

I initially only had the config loaded and managed in the main process with IPC functions available for the renderer process to use to retrieve and set config. Unfortunately that caused any test using config to fail as the simulated `config.json` files used in testing didn't seem to be available to the main process and so the renderer process could never get the config it was expecting. The quick solution was to load an instance of the same config class in both the main and renderer processes to be used directly and then add functionality to keep them in sync.

There is probably a better solution given enough time and if I can get debugging in tests to work, but I took this approach because we need it for MSI/GPO.

Does that make sense?

Reply:
yup it does. thanks for the explanation